File: day_06.go
package main

import "fmt"

func isStartOfPacketMarker(s string) bool {
	seenChars := make(map[byte]bool)

	for i := 0; i < len(s); i++ {
		if seenChars[s[i]] {
			return false
		} else {
			seenChars[s[i]] = true
		}
	}

	return true
}

func nCharsNeededToReceiveUniqueMarker(datastreamBuffer string, markerLength int) int {
	for i := 0; i <= len(datastreamBuffer)-markerLength; i++ {
		if isStartOfPacketMarker(datastreamBuffer[i : i+markerLength]) {
			return i + markerLength
		}
	}
	return 0
}

func Day06() {
	datastreamBuffer := ReadLines("data/day_06.txt")[0]

	fmt.Println("part 1", nCharsNeededToReceiveUniqueMarker(datastreamBuffer, 4))
	fmt.Println("part 2", nCharsNeededToReceiveUniqueMarker(datastreamBuffer, 14))
}

File: day_06.rb
# frozen_string_literal: true

def start_of_packet_marker?(string)
  string.chars.uniq.count == string.length
end

def n_chars_needed_to_receive_unique_marker(datastream_buffer, marker_length:)
  0..(datastream_buffer.length - (marker_length - 1)).times do |index|
    return index + marker_length if start_of_packet_marker?(datastream_buffer[index, marker_length])
  end
end

if __FILE__ == $PROGRAM_NAME
  input_filepath = 'data/day_06.txt'
  datastream_buffer = File.read(input_filepath)

  puts n_chars_needed_to_receive_unique_marker(datastream_buffer, marker_length: 4)
  puts n_chars_needed_to_receive_unique_marker(datastream_buffer, marker_length: 14)
end
